port/linux/omrvmem.c

```diff
@@ -1076,61 +1094,91 @@ get_hugepages_info(struct OMRPortLibrary *portLibrary, vmem_hugepage_info_t *pag
+#else
+	useBackingFile = TRUE;
+	*flags |= MAP_PRIVATE;
+#endif
+	return useBackingFile;
+}
+
 static void *
-default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *address, uintptr_t byteAmount, struct J9PortVmemIdentifier *identifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category)
+reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uintptr_t byteAmount, struct J9PortVmemIdentifier *identifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category)
 {
 	/* This function is cloned in J9SourceUnixJ9VMem (omrvmem_reserve_memory).
 	 * Any changes made here may need to be reflected in that version.
 	 */
 	int fd = -1;
-	int ft = -1;
 	int flags = 0;
 	void *result = NULL;
 	int protectionFlags = PROT_NONE;
-	BOOLEAN useBackingSharedFile = FALSE;
 	BOOLEAN useBackingFile = FALSE;
+	BOOLEAN useBackingSharedFile = FALSE;
+	BOOLEAN isHugePagesEnabled = FALSE;
 
-	Trc_PRT_vmem_default_reserve_entry(address, byteAmount);
+	Trc_PRT_vmem_reserve_entry(address, byteAmount);
 
-	if(mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) {
+	if (mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) {
 		flags |= MAP_SHARED;
 		useBackingSharedFile = TRUE;
 	} else {
-#if defined(MAP_ANONYMOUS)
-		flags |= MAP_ANONYMOUS | MAP_PRIVATE;
-		useBackingSharedFile = FALSE;
-#elif defined(MAP_ANON)
-		flags |= MAP_ANON | MAP_PRIVATE;
-		useBackingSharedFile = FALSE;
-#else
-		useBackingFile = TRUE;
-		flags |= MAP_PRIVATE;
-#endif
+		useBackingFile = set_flags_for_mmap(&flags);
+	}
+
+#if defined(MAP_HUGETLB)
+	if (0 != (mode & OMRPORT_VMEM_MEMORY_MODE_MMAP_HUGE_PAGES)) {
+		flags |= MAP_HUGETLB;
+		isHugePagesEnabled = TRUE;
 	}
+#endif /* MAP_HUGETLB */
 
 	if (useBackingSharedFile) {
-		char filename[FILE_NAME_SIZE + 1];
-		sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
+		memfd_call memfd_createDL = (memfd_call)portLibrary->portGlobals->memfd_createDL;
+
+		if (NULL != memfd_createDL) {
+			Trc_PRT_vmem_reserve_doubleMapAPI_available(address, byteAmount);
+			mode |= OMRPORT_VMEM_MEMORY_MODE_DOUBLE_MAP_AVAILABLE;
+			int ft = -1;
+			uintptr_t fdFlags = 0;
+			char filename[FILE_NAME_SIZE + 1];
+			sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  (size_t)omrtime_current_time_millis(portLibrary),
 								getpid(),
 								(size_t)pthread_self());
-		fd = shm_open(filename, O_RDWR | O_CREAT, 0600);
-		shm_unlink(filename);
-		ft = ftruncate(fd, byteAmount);
-
-		if (fd == -1 || ft == -1) {
-			Trc_PRT_vmem_default_reserve_failed(address, byteAmount);
-			update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
-			Trc_PRT_vmem_default_reserve_exit(result, address, byteAmount);
-			return result;
+#if defined(MFD_HUGETLB)
+			if (isHugePagesEnabled) {
+				fdFlags = MFD_HUGETLB;
+			}
+#endif /* defined(MFD_HUGETLB) */
+			fd = memfd_createDL(filename, fdFlags);
+			ft = ftruncate(fd, byteAmount);
+
+			if (fd == -1 || ft == -1) {
```

Comment:
technically would be better to have constant first as well as extra brackets `if ((-1 == fd) || (-1 == ft))`